File: Stack/1209. Remove All Adjacent Duplicates in String II.java
class Solution {
    public String removeDuplicates(String s, int k) {
        if (s == null || s.length() == 0) {
            return new String();
        }
        Stack<Character> stackChar = new Stack<>();
        Stack<Integer> stackInt = new Stack<>();
        for (char c : s.toCharArray()) {
            if (!stackChar.isEmpty() && stackChar.peek() == c) {
                int times = stackInt.pop();
                if (times + 1 == k) {
                    stackChar.pop();
                } else {
                    stackInt.push(times + 1);
                }
            } else {
                stackChar.push(c);
                stackInt.push(1);
            }
        }
        // System.out.println(stackChar);
        // System.out.println(stackInt);

        StringBuilder sb = new StringBuilder();
        while (!stackChar.isEmpty()) {
            char c = stackChar.pop();
            int times = stackInt.pop();
            String str = Character.toString(c);
            sb.append(str.repeat(times));
        }
        return sb.reverse().toString();
    }
}

File: Stack/1209. Remove All Adjacent Duplicates in String II.py
class Solution:
    def removeDuplicates(self, s: str, k: int) -> str:
        # stack store (char, times)
        # if times + 1 == k:
        #   pop
        # [(d,1), (e,3)]
        # time O(n), space O(n)
        stack = []
        for char in s:
            if stack and stack[-1][0] == char:
                _, times = stack.pop()
                if times + 1 < k:
                    stack.append((char, times + 1))
            else:
                stack.append((char, 1))
        res = ''
        for char, t in stack:
            res += char * t
        return res
